ref(discover): Cleaning up how we build the filter params

wmak opened the pull request:
- This generally adds a max fields check, and multiple project
  permission checks and date quantizing to all the discover endpoints
- I'm most fuzzy about the new permission check, it specifically means
  that orgs without "organizations:global-views" will no longer be able
  to use the following endpoints on multiple projects:
  - tags/values
  - events-stats
  - events-facets
  - events-meta
  - event-baseline
  - related-issues
- Also minor change to trends, to not get params twice

wmak committed:
ref(discover): Cleaning up how we build the filter params

- This generally adds a max fields check, and multiple project
  permission checks and date quantizing to all the discover endpoints
- I'm most fuzzy about the new permission checks, it specifically means
  that orgs without "organizations:global-views" will no longer be able
  to see the following on multiple projects:
  - tags/values
  - events-stats
  - events-facets
  - events-meta
  - event-baseline
  - related-issues

markstory commented on src/sentry/api/bases/organization_events.py:
```diff
@@ -33,13 +34,34 @@ def has_feature(self, organization, request):
+                    )
+                )
+
+            span.set_tag("organization", organization)
```
If the organization is already an event level tag do we need it on the span as well?

markstory commented on src/sentry/api/bases/organization_events.py:
```diff
@@ -194,19 +216,26 @@ def handle_data(self, request, organization, project_ids, results):
+                    try:
+                        params = self.get_snuba_params(request, organization)
+                    except NoProjects:
+                        return {"data": []}
```
Are we able to update all the callers so that this code path isn't required?

wmak replied:
Unfortunately not, some of the endpoints act differently when there are no projects. Eg. related-issues returns an empty list, but here we return a dictionary instead.

markstory replied:
That's unfortunate.

markstory commented on tests/snuba/api/endpoints/test_organization_events_stats.py:
```diff
@@ -63,18 +63,33 @@ def setUp(self):
         )
 
     def test_simple(self):
+        with self.feature("organizations:global-views"):
```
Do the old events views still work with the new feature flags applied? Those accounts won't have access to new discover.

wmak replied:
Oof ty, I knew there must have been a reason why this endpoint didn't check global views. Checked and removed the global views check on `events-stats`, `events-meta` and tag value keys which are all still being used by the legacy events view.

Zylphrex commented on src/sentry/api/bases/organization_events.py:
```diff
@@ -33,13 +34,34 @@ def has_feature(self, organization, request):
+                "organizations:global-views", organization, actor=request.user
+            )
+            if not has_global_views and len(params.get("project_id", [])) > 1:
+                raise ParseError(detail="You cannot view events from multiple projects.")
```
I know this is a refactor but shouldn't this be a 404?

wmak replied:
🤔 should it be a 404? At this point we've checked that the user can use the endpoint that we're getting params for, but what they're asking for is invalid

Zylphrex replied:
I'd argue that this is a permission error as they don't have the right features enabled to do this. The following snippet is in many parts of the code base

```python
if not self.has_feature(organization, request):
    return Response(status=404) 
```

And I'd argue the error in the two cases should at least be the same. Though I do see what you mean.

wmak replied:
@Zylphrex Hmm, I think that's a reasonable argument too. But to counter that, other parts of the app 400 on this error too (eg. issues). And I think part of the reason for that is that with a 400 we can include the detail to give a helpful error message.

Zylphrex commented on src/sentry/api/endpoints/organization_events_meta.py:
```diff
@@ -101,12 +95,13 @@ def get(self, request, organization):
+            return Response([])
+
+        with sentry_sdk.start_span(op="discover.endpoint", description="find_lookup_keys") as span:
             span.set_data("organization", organization)
```
This uses `set_data` rather than `set_tag` to attach the organization. Should this be unified?

wmak replied:
Mark's comment [here](https://github.com/getsentry/sentry/pull/20849/files#r490976642) raises a good point, I'm just going to remove this instead since we already have the org at event level

wmak committed:
ref: Adding a param to check global views

- Events stats and events meta are both shared with the legacy events
  view. Which allows organizations without global-views to still check
  events across multiple projects
- Also removed the redundant organization set_tag

wmak committed:
Merge branch 'master' into ref/consolidate-filter-params
